packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -2628,10 +2675,33 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
     }
     assert(!_debugLocked);
     assert(() { _debugLocked = true; return true; }());
+    // We have to manually extract the inherited widget in initState because
+    // the current context is not fully initialized.
+    final InheritedNavigatorObserver inheritedNavigatorObserver = context
+      .getElementForInheritedWidgetOfExactType<InheritedNavigatorObserver>()
+      ?.widget as InheritedNavigatorObserver;
+    _updateInheritedObserver(inheritedNavigatorObserver?.observer);
     _flushHistoryUpdates();
+    // Reset the reference so that it can be initialized in the
+    // didChangeDependencies.
+    _updateInheritedObserver(null);
```

Comment:
Why do we need to do this? if `didChangeDependencies` finds a different InheritedNavigatorObserver it would just update it. If it finds the same `_updateInheritedObserver` would just short circuit.

Reply:
Ah you are right the InheritedNavigatorObserver.of will get called in didChangeDependencies anyway even if it will short circuit in _updateInheritedObserver.